apps/emqx_gcp_device/src/emqx_gcp_device.erl

```diff
@@ -0,0 +1,256 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2023 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+
+-module(emqx_gcp_device).
+
+-include_lib("emqx_authn/include/emqx_authn.hrl").
+-include_lib("emqx/include/emqx.hrl").
+-include_lib("emqx/include/logger.hrl").
+-include_lib("stdlib/include/ms_transform.hrl").
+
+%% Management
+-export([put_device/1, get_device/1, remove_device/1]).
+%% Management: import
+-export([import_devices/1]).
+%% Authentication
+-export([get_device_actual_keys/1]).
+%% Internal API
+-export([create_table/0, clear_table/0, format_device/1, config_topic/1]).
+
+-type deviceid() :: binary().
+-type project() :: binary().
+-type location() :: binary().
+-type registry() :: binary().
+-type device_loc() :: {project(), location(), registry()}.
+-type key_type() :: binary().
+-type key() :: binary().
+-type expires_at() :: pos_integer().
+-type key_record() :: {key_type(), key(), expires_at()}.
+-type created_at() :: pos_integer().
+-type extra() :: map().
+
+-record(emqx_gcp_device, {
+    id :: deviceid(),
+    keys :: [key_record()],
+    device_loc :: device_loc(),
+    created_at :: created_at(),
+    extra :: extra()
+}).
+-type emqx_gcp_device() :: #emqx_gcp_device{}.
+
+-type formatted_key() ::
+    #{
+        key_type := key_type(),
+        key := key(),
+        expires_at := expires_at()
+    }.
+-type encoded_config() :: binary().
+-type formatted_device() ::
+    #{
+        deviceid := deviceid(),
+        keys := [formatted_key()],
+        config := encoded_config(),
+        project => project(),
+        location => location(),
+        registry => registry(),
+        created_at => created_at()
+    }.
+-export_type([formatted_device/0, deviceid/0, encoded_config/0]).
+
+-define(TAB, ?MODULE).
+
+%%--------------------------------------------------------------------
+%% API
+%%--------------------------------------------------------------------
+
+-spec put_device(formatted_device()) -> ok.
+put_device(FormattedDevice) ->
+    try
+        transaction(fun() -> put_device_no_transaction(FormattedDevice) end)
+    catch
+        _Error:Reason ->
+            ?SLOG(error, #{
+                msg => "Failed to put device",
+                device => FormattedDevice,
+                reason => Reason
+            }),
+            {error, Reason}
+    end.
+
+-spec get_device(deviceid()) -> {ok, formatted_device()} | not_found.
+get_device(DeviceId) ->
+    try ets:lookup(?TAB, DeviceId) of
+        [] ->
+            not_found;
+        [Device] ->
+            {ok, format_device(Device)}
+    catch
+        error:badarg ->
```

Comment:
I think in a normal situation, this clause should never be reached.

Reply:
I'll remove the catch